components/automate-gateway/gateway/server.go

```diff
@@ -325,42 +125,314 @@ func (s *Server) NewGRPCServer() (*grpc.Server, error) {
+	s.loadConnFactory()
+	s.loadServiceCerts()
+
+	err := s.startNullBackendServer()
+	if err != nil {
+		return errors.Wrap(err, "starting null backend")
+	}
+
+	err = s.loadClients()
+	if err != nil {
+		return errors.Wrap(err, "loading backend gRPC clients")
+	}
+
+	err = s.loadAuthorizer()
+	if err != nil {
+		return errors.Wrap(err, "loading authorizor")
+	}
+
+	err = s.startGRPCServer()
+	if err != nil {
+		return errors.Wrap(err, "starting gateway gRPC server")
+	}
+
+	err = s.startHTTPServer()
+	if err != nil {
+		return errors.Wrap(err, "starting gateway HTTPS server")
+	}
+
+	return s.startSignalHandler()
+}
+
+func (s *Server) loadConnFactory() {
+	s.logger.Debug("loading gRPC connection factory")
+	s.connFactory = secureconn.NewFactory(*s.Config.ServiceCerts, secureconn.DisableDebugServer())
+
+	return
+}
+
+func (s *Server) loadClients() error {
+	var err error
+
+	s.logger.Info("loading backend gRPC clients")
+
+	s.clientsFactory, err = NewClientsFactory(s.Config.GrpcClients, s.connFactory)
+
+	return err
+}
+
+func (s *Server) loadServiceCerts() {
+	s.logger.Debug("loading automate-gateway service certs")
+
+	s.serviceKeyPair = s.Config.ServiceCerts.ServiceKeyPair
+	s.rootCerts = s.Config.ServiceCerts.NewCertPool()
+
+	return
+}
+
+func (s *Server) setLogLevel() {
+	l, err := log.ParseLevel(s.Config.Log.Level)
+	if err != nil {
+		log.Warnf("unknown log level %q, using default (info)", l)
+		l = log.InfoLevel
+	}
+	log.SetLevel(l)
+
+	return
+}
+
+func (s *Server) loadAuthorizer() error {
+	s.logger.Info("loading authorizer")
+
+	authzClientV1, err := s.clientsFactory.AuthorizationClient()
+	if err != nil {
+		return errors.Wrap(err, "create authz client")
+	}
+	authzClientV2, err := s.clientsFactory.AuthorizationV2Client()
+	if err != nil {
+		return errors.Wrap(err, "create authz_v2 client")
+	}
+
+	s.authorizer = authorizer.NewAuthorizer(
+		authv1.AuthorizationHandler(authzClientV1),
+		authv2.AuthorizationHandler(authzClientV2),
+	)
+
+	return nil
+}
+
+// startNullBackendServer starts the unimplemented backend server
+func (s *Server) startNullBackendServer() error {
+	var err error
+
+	s.logger.Info("starting null backend server")
+
+	nullBackendListener, err := net.Listen("unix", s.Config.GrpcClients.NullBackendSock)
+	if err != nil {
+		return errors.Wrapf(err, "listen on %s", s.Config.GrpcClients.NullBackendSock)
+	}
+
+	s.nullBackendServer = nullbackend.NewServer()
+
+	// Start all servers in goroutines and start watching for user signals. If
+	// an error is returned from a server or an exit signal is received, try to
+	// gracefully stop servers and exit.
+	go func() {
+		err := s.nullBackendServer.Serve(nullBackendListener)
+		if err != nil {
+			s.errC <- errors.Wrap(err, "serve null backend")
+		}
+		return
+	}()
+
+	return nil
+}
+
+func (s *Server) stopNullBackendServer() {
+	s.logger.Info("stopping null backend server")
+
+	s.nullBackendServer.GracefulStop()
+	_ = os.Remove(s.Config.GrpcClients.NullBackendSock)
```

Comment:
what happens if this file is not removed

Reply:
Nothing really, because the hab run hook will delete it during start-up if the old socket gets left around. Here we're just being good stewards and cleaning up after ourselves.